encoders/firebase-encoders-reflective/src/main/java/com/google/firebase/encoders/reflective/InternalAnnotationContext.java

```diff
@@ -0,0 +1,143 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.encoders.reflective;
+
+import com.google.firebase.encoders.EncodingException;
+import com.google.firebase.encoders.annotations.Encodable;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+final class InternalAnnotationContext {
+  private boolean inline;
+  private String decodingKey;
+  private boolean ignored;
+
+  InternalAnnotationContext(AccessibleObject accObj) {
+    if (accObj instanceof Field) {
+      decodingKey = ((Field) accObj).getName();
+    } else if (accObj instanceof Method) {
+      decodingKey = fieldName((Method) accObj);
+    } else {
+      throw new EncodingException("Constructor shouldn't be used to get its decoding key");
+    }
+    this.inline = false;
+    this.ignored = false;
+
+    // internal annotations
+    if (accObj.isAnnotationPresent(Encodable.Ignore.class)) {
+      this.setIgnored(true);
+      return;
+    }
+    if (accObj.isAnnotationPresent(Encodable.Field.class)) {
+      Encodable.Field annotation = accObj.getAnnotation(Encodable.Field.class);
+      if (annotation != null && annotation.name().length() > 0) {
+        this.setDecodingKey(annotation.name());
+      }
+      if (annotation != null) {
+        this.setInline(annotation.inline());
+      }
+    }
+
+    // alias annotations
+    for (Annotation annotation : accObj.getAnnotations()) {
+      Alias alias = annotation.annotationType().getAnnotation(Alias.class);
+      if (alias != null) {
+        Class<? extends Annotation> actualAnnotationType = alias.value();
+        if (actualAnnotationType.equals(Encodable.Ignore.class)) {
+          this.setIgnored(true);
+        } else if (actualAnnotationType.equals(Encodable.Field.class)) {
+          for (Method method : annotation.annotationType().getDeclaredMethods()) {
+            Alias.Property property = method.getAnnotation(Alias.Property.class);
+            if (property == null) {
+              continue;
+            }
+            String propertyName = property.value();
+            Object obj = null;
+            try {
+              method.setAccessible(true);
+              obj = method.invoke(annotation);
+            } catch (IllegalAccessException e) {
+              throw new RuntimeException(e);
+            } catch (InvocationTargetException e) {
+              throw new RuntimeException(e);
+            }
+            if (propertyName.equals("name")) {
+              String val = (String) obj;
```

Comment:
This cast could fail if my annotation has a different type for the property, i.e.

```java
@Alias
@interface Foo {
  @Alias.Property("name")
  int foo();
}
```
We should throw an `EncodingException` with a descriptive message.